### ISSUE TITLE
Some problems about training

### ISSUE THREAD

**Issue by Iwill-github** [ultralytics/ultralytics#6223]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, contributors. Recently, I used batchsize=128 and epoch=500 to reproduce yolov8s. The accuracy is only 44.1, which is lower than 44.9. Did you use multi-scale training or other data enhancement strategies during your training?

**Comment by glenn-jocher** [comment]
@Iwill-github hello,

Thank you for reaching out and for conducting a search before posting your question. When training YOLOv8 models to reproduce results, there are several factors that can impact the final accuracy.

Firstly, yes, during training, YOLOv8 does leverage multi-scale training as well as a variety of data augmentation strategies to improve robustness and performance. The variances in training parameters, data preprocessing, and augmentation can lead to differences in the final accuracy when reproducing the models.

Regarding the batch size and epochs you mentioned, it’s important to ensure that you are following the same setup as reported in the official experiments, as any deviations might affect the outcome. Additionally, the learning rate and its schedule, weight decay, and other hyperparameters should closely match the original training configuration.

It's also worth noting that different random seeds and minor differences in the training environment could lead to variability in the results. The accuracy can fluctuate slightly with every training due to the stochastic nature of deep learning models.

To get closer to the reported accuracy, consider verifying that your setup aligns with the original YOLOv8s configurations. Also, ensure your environment is consistent with the one used to obtain the reported results, paying attention to software versions and hardware specifications. Lastly, running multiple training cycles and averaging the results can give a more stable estimate of the model’s performance.

If after reviewing your setup, you don't observe any disparities, and you continue to encounter the same issue, please provide us with details regarding your training configuration, environment, and any modifications you've made. This information will greatly assist us in providing more specific guidance.

Best regards,
Glenn Jocher

**Comment by Iwill-github** [comment]
First of all, thank you very much for your patient reply.

I used two RTX 3090 for training in the work of reproducing yolov8s, and used the yaml file shown below. The changed parameters have been listed in the first few rows, and the rest of the superparameters follow the default.yaml.

I wonder if I didn't achieve your results because I didn't use mixup or retina_masks? Also, I didn't find the setting of multi-scale training in the configuration file. Is it the default to use multi-scale training?

Finally, if possible, could you please give some suggestions for improving the super-parameter configuration?



`
# modified parameters -------------------------------------------------------------------------------------------------------
epochs: 500  # (int) number of epochs to train for
batch: 128  # (int) number of images per batch (-1 for AutoBatch)
imgsz: 640  # (int | list) input images size as int for train and val modes, or list[w,h] for predict and export modes
save_json: True  # (bool) save results to JSON file


# Ultralytics YOLO 🚀, AGPL-3.0 license
# Default training settings and hyperparameters for medium-augmentation COCO training

task: detect  # (str) YOLO task, i.e. detect, segment, classify, pose
mode: train  # (str) YOLO mode, i.e. train, val, predict, export, track, benchmark

# Train settings -------------------------------------------------------------------------------------------------------
model:  # (str, optional) path to model file, i.e. yolov8n.pt, yolov8n.yaml
data:  # (str, optional) path to data file, i.e. coco128.yaml
patience: 50  # (int) epochs to wait for no observable improvement for early stopping of training
save: True  # (bool) save train checkpoints and predict results
save_period: -1 # (int) Save checkpoint every x epochs (disabled if < 1)
cache: False  # (bool) True/ram, disk or False. Use cache for data loading
device:  # (int | str | list, optional) device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8  # (int) number of worker threads for data loading (per RANK if DDP)
project:  # (str, optional) project name
name:     # (str, optional) experiment name, results saved to 'project/name' directory
exist_ok: False  # (bool) whether to overwrite existing experiment
pretrained: True  # (bool | str) whether to use a pretrained model (bool) or a model to load weights from (str)
optimizer: auto  # (str) optimizer to use, choices=[SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, auto]
verbose: True  # (bool) whether to print verbose output
seed: 0  # (int) random seed for reproducibility
deterministic: True  # (bool) whether to enable deterministic mode
single_cls: False  # (bool) train multi-class data as single-class
rect: False  # (bool) rectangular training if mode='train' or rectangular validation if mode='val'
cos_lr: False  # (bool) use cosine learning rate scheduler
close_mosaic: 10  # (int) disable mosaic augmentation for final epochs (0 to disable)
resume: False  # (bool) resume training from last checkpoint
amp: True  # (bool) Automatic Mixed Precision (AMP) training, choices=[True, False], True runs AMP check
fraction: 1.0  # (float) dataset fraction to train on (default is 1.0, all images in train set)
profile: False  # (bool) profile ONNX and TensorRT speeds during training for loggers
freeze: None  # (int | list, optional) freeze first n layers, or freeze list of layer indices during training
# Segmentation
overlap_mask: True  # (bool) masks should overlap during training (segment train only)
mask_ratio: 4  # (int) mask downsample ratio (segment train only)
# Classification
dropout: 0.0  # (float) use dropout regularization (classify train only)

# Val/Test settings ----------------------------------------------------------------------------------------------------
val: True  # (bool) validate/test during training
split: val  # (str) dataset split to use for validation, i.e. 'val', 'test' or 'train'
save_hybrid: False  # (bool) save hybrid version of labels (labels + additional predictions)
conf:  # (float, optional) object confidence threshold for detection (default 0.25 predict, 0.001 val)
iou: 0.7  # (float) intersection over union (IoU) threshold for NMS
max_det: 300  # (int) maximum number of detections per image
half: False  # (bool) use half precision (FP16)
dnn: False  # (bool) use OpenCV DNN for ONNX inference
plots: True  # (bool) save plots during train/val

# Prediction settings --------------------------------------------------------------------------------------------------
source:  # (str, optional) source directory for images or videos
show: False  # (bool) show results if possible
save_txt: False  # (bool) save results as .txt file
save_conf: False  # (bool) save results with confidence scores
save_crop: False  # (bool) save cropped images with results
show_labels: True  # (bool) show object labels in plots
show_conf: True  # (bool) show object confidence scores in plots
vid_stride: 1  # (int) video frame-rate stride
line_width:   # (int, optional) line width of the bounding boxes, auto if missing
visualize: False  # (bool) visualize model features
augment: False  # (bool) apply image augmentation to prediction sources
agnostic_nms: False  # (bool) class-agnostic NMS
classes:  # (int | list[int], optional) filter results by class, i.e. classes=0, or classes=[0,2,3]
retina_masks: False  # (bool) use high-resolution segmentation masks
boxes: True  # (bool) Show boxes in segmentation predictions

# Export settings ------------------------------------------------------------------------------------------------------
format: torchscript  # (str) format to export to, choices at https://docs.ultralytics.com/modes/export/#export-formats
keras: False  # (bool) use Kera=s
optimize: False  # (bool) TorchScript: optimize for mobile
int8: False  # (bool) CoreML/TF INT8 quantization
dynamic: False  # (bool) ONNX/TF/TensorRT: dynamic axes
simplify: False  # (bool) ONNX: simplify model
opset:  # (int, optional) ONNX: opset version
workspace: 4  # (int) TensorRT: workspace size (GB)
nms: False  # (bool) CoreML: add NMS

# Hyperparameters ------------------------------------------------------------------------------------------------------
lr0: 0.01  # (float) initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.01  # (float) final learning rate (lr0 * lrf)
momentum: 0.937  # (float) SGD momentum/Adam beta1
weight_decay: 0.0005  # (float) optimizer weight decay 5e-4
warmup_epochs: 3.0  # (float) warmup epochs (fractions ok)
warmup_momentum: 0.8  # (float) warmup initial momentum
warmup_bias_lr: 0.1  # (float) warmup initial bias lr
box: 7.5  # (float) box loss gain
cls: 0.5  # (float) cls loss gain (scale with pixels)
dfl: 1.5  # (float) dfl loss gain
pose: 12.0  # (float) pose loss gain
kobj: 1.0  # (float) keypoint obj loss gain
label_smoothing: 0.0  # (float) label smoothing (fraction)
nbs: 64  # (int) nominal batch size
hsv_h: 0.015  # (float) image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # (float) image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # (float) image HSV-Value augmentation (fraction)
degrees: 0.0  # (float) image rotation (+/- deg)
translate: 0.1  # (float) image translation (+/- fraction)
scale: 0.5  # (float) image scale (+/- gain)
shear: 0.0  # (float) image shear (+/- deg)
perspective: 0.0  # (float) image perspective (+/- fraction), range 0-0.001
flipud: 0.0  # (float) image flip up-down (probability)
fliplr: 0.5  # (float) image flip left-right (probability)
mosaic: 1.0  # (float) image mosaic (probability)
mixup: 0.0  # (float) image mixup (probability)
copy_paste: 0.0  # (float) segment copy-paste (probability)

# Custom config.yaml ---------------------------------------------------------------------------------------------------
cfg:  # (str, optional) for overriding defaults.yaml

# Tracker settings ------------------------------------------------------------------------------------------------------
tracker: botsort.yaml  # (str) tracker type, choices=[botsort.yaml, bytetrack.yaml]
`

**Comment by Iwill-github** [comment]
First of all, thank you very much for your patient reply.

I used two RTX3090s for training in the work of reproducing yolov8s, and used the yaml settings shown below. The changed parameters have been listed in the first few rows, and the rest of the superparameters follow the default.yaml.

I wonder if I didn't achieve your results because I didn't use mixup or retina_masks? Also, I didn't find the setting of multi-scale training in the configuration file. Is it the default to use multi-scale training?

Finally, if possible, could you please give some suggestions for improving the super-parameter configuration?







# modified parameters -------------------------------------------------------------------------------------------------------
epochs: 500  # (int) number of epochs to train for
batch: 128  # (int) number of images per batch (-1 for AutoBatch)
imgsz: 640  # (int | list) input images size as int for train and val modes, or list[w,h] for predict and export modes
save_json: True  # (bool) save results to JSON file




# Ultralytics YOLO 🚀, AGPL-3.0 license
# Default training settings and hyperparameters for medium-augmentation COCO training

task: detect  # (str) YOLO task, i.e. detect, segment, classify, pose
mode: train  # (str) YOLO mode, i.e. train, val, predict, export, track, benchmark



# Train settings -------------------------------------------------------------------------------------------------------
model:  # (str, optional) path to model file, i.e. yolov8n.pt, yolov8n.yaml
data:  # (str, optional) path to data file, i.e. coco128.yaml
patience: 50  # (int) epochs to wait for no observable improvement for early stopping of training
save: True  # (bool) save train checkpoints and predict results
save_period: -1 # (int) Save checkpoint every x epochs (disabled if < 1)
cache: False  # (bool) True/ram, disk or False. Use cache for data loading
device:  # (int | str | list, optional) device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8  # (int) number of worker threads for data loading (per RANK if DDP)
project:  # (str, optional) project name
name:     # (str, optional) experiment name, results saved to 'project/name' directory
exist_ok: False  # (bool) whether to overwrite existing experiment
pretrained: True  # (bool | str) whether to use a pretrained model (bool) or a model to load weights from (str)
optimizer: auto  # (str) optimizer to use, choices=[SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, auto]
verbose: True  # (bool) whether to print verbose output
seed: 0  # (int) random seed for reproducibility
deterministic: True  # (bool) whether to enable deterministic mode
single_cls: False  # (bool) train multi-class data as single-class
rect: False  # (bool) rectangular training if mode='train' or rectangular validation if mode='val'
cos_lr: False  # (bool) use cosine learning rate scheduler
close_mosaic: 10  # (int) disable mosaic augmentation for final epochs (0 to disable)
resume: False  # (bool) resume training from last checkpoint
amp: True  # (bool) Automatic Mixed Precision (AMP) training, choices=[True, False], True runs AMP check
fraction: 1.0  # (float) dataset fraction to train on (default is 1.0, all images in train set)
profile: False  # (bool) profile ONNX and TensorRT speeds during training for loggers
freeze: None  # (int | list, optional) freeze first n layers, or freeze list of layer indices during training
# Segmentation
overlap_mask: True  # (bool) masks should overlap during training (segment train only)
mask_ratio: 4  # (int) mask downsample ratio (segment train only)
# Classification
dropout: 0.0  # (float) use dropout regularization (classify train only)

# Val/Test settings ----------------------------------------------------------------------------------------------------
val: True  # (bool) validate/test during training
split: val  # (str) dataset split to use for validation, i.e. 'val', 'test' or 'train'
save_hybrid: False  # (bool) save hybrid version of labels (labels + additional predictions)
conf:  # (float, optional) object confidence threshold for detection (default 0.25 predict, 0.001 val)
iou: 0.7  # (float) intersection over union (IoU) threshold for NMS
max_det: 300  # (int) maximum number of detections per image
half: False  # (bool) use half precision (FP16)
dnn: False  # (bool) use OpenCV DNN for ONNX inference
plots: True  # (bool) save plots during train/val

# Prediction settings --------------------------------------------------------------------------------------------------
source:  # (str, optional) source directory for images or videos
show: False  # (bool) show results if possible
save_txt: False  # (bool) save results as .txt file
save_conf: False  # (bool) save results with confidence scores
save_crop: False  # (bool) save cropped images with results
show_labels: True  # (bool) show object labels in plots
show_conf: True  # (bool) show object confidence scores in plots
vid_stride: 1  # (int) video frame-rate stride
line_width:   # (int, optional) line width of the bounding boxes, auto if missing
visualize: False  # (bool) visualize model features
augment: False  # (bool) apply image augmentation to prediction sources
agnostic_nms: False  # (bool) class-agnostic NMS
classes:  # (int | list[int], optional) filter results by class, i.e. classes=0, or classes=[0,2,3]
retina_masks: False  # (bool) use high-resolution segmentation masks
boxes: True  # (bool) Show boxes in segmentation predictions

# Export settings ------------------------------------------------------------------------------------------------------
format: torchscript  # (str) format to export to, choices at https://docs.ultralytics.com/modes/export/#export-formats
keras: False  # (bool) use Kera=s
optimize: False  # (bool) TorchScript: optimize for mobile
int8: False  # (bool) CoreML/TF INT8 quantization
dynamic: False  # (bool) ONNX/TF/TensorRT: dynamic axes
simplify: False  # (bool) ONNX: simplify model
opset:  # (int, optional) ONNX: opset version
workspace: 4  # (int) TensorRT: workspace size (GB)
nms: False  # (bool) CoreML: add NMS

# Hyperparameters ------------------------------------------------------------------------------------------------------
lr0: 0.01  # (float) initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.01  # (float) final learning rate (lr0 * lrf)
momentum: 0.937  # (float) SGD momentum/Adam beta1
weight_decay: 0.0005  # (float) optimizer weight decay 5e-4
warmup_epochs: 3.0  # (float) warmup epochs (fractions ok)
warmup_momentum: 0.8  # (float) warmup initial momentum
warmup_bias_lr: 0.1  # (float) warmup initial bias lr
box: 7.5  # (float) box loss gain
cls: 0.5  # (float) cls loss gain (scale with pixels)
dfl: 1.5  # (float) dfl loss gain
pose: 12.0  # (float) pose loss gain
kobj: 1.0  # (float) keypoint obj loss gain
label_smoothing: 0.0  # (float) label smoothing (fraction)
nbs: 64  # (int) nominal batch size
hsv_h: 0.015  # (float) image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # (float) image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # (float) image HSV-Value augmentation (fraction)
degrees: 0.0  # (float) image rotation (+/- deg)
translate: 0.1  # (float) image translation (+/- fraction)
scale: 0.5  # (float) image scale (+/- gain)
shear: 0.0  # (float) image shear (+/- deg)
perspective: 0.0  # (float) image perspective (+/- fraction), range 0-0.001
flipud: 0.0  # (float) image flip up-down (probability)
fliplr: 0.5  # (float) image flip left-right (probability)
mosaic: 1.0  # (float) image mosaic (probability)
mixup: 0.0  # (float) image mixup (probability)
copy_paste: 0.0  # (float) segment copy-paste (probability)

# Custom config.yaml ---------------------------------------------------------------------------------------------------
cfg:  # (str, optional) for overriding defaults.yaml

# Tracker settings ------------------------------------------------------------------------------------------------------
tracker: botsort.yaml  # (str) tracker type, choices=[botsort.yaml, bytetrack.yaml]

**Comment by glenn-jocher** [comment]
@Iwill-github, it’s great to see the details of your configuration. Let’s walk through some of the points you've raised.

Multi-scale training: This is indeed a default behavior in YOLOv8 and is managed internally by the training loop. The model is trained with varying image sizes within a predefined range every certain number of batches.

Mixup and Data Augmentation: ‘Mixup’ and other data augmentation techniques like 'mosaic', 'cutmix', and 'copy_paste' can significantly impact the performance. If these settings were used in the official training but not in your reproduction, this might contribute to the performance delta you're observing. The probabilities for these augmentations, when set to values other than 0.0, indicate that they're active during training.

Retina Masks: The `retina_masks` setting in the prediction section doesn’t directly impact training, but rather the inference when using segmentation masks.

To improve your super-parameter configuration:

1. Data Augmentation: Consider incorporating more data augmentation techniques if the model has been officially trained using them. For instance, set 'mixup', 'mosaic', and 'copy_paste' augmentation probabilities to values greater than 0.0.

2. Learning Rate Scheduler: Verify if a learning rate schedule was in place, which could involve a 'cos_lr' policy or a custom scheduler, as it can significantly impact final performance.

3. Batch Size and Synchronization: With two GPUs, ensure batch size synchronization across GPUs. When using Distributed Data Parallel (DDP), the effective batch size would be double the batch size you specify, as gradients are summed across GPUs.

4. Image Sizes: Double-check if the chosen image size (640) is consistent with the model's original training configuration.

5. Epochs and Early Stopping: While you have specified 500 epochs, the patience parameter also plays a role. If you see that the model is not improving over a span of 50 epochs (as you have set 'patience'), adjust this parameter if necessary.

6. Hyperparameters: Delicate tuning of learning rate (lr0), weight decay, and other hyperparameters included in the configuration file can lead to better convergence and accuracy.

7. If none of this helps, investigate your dataset and training environment for other discrepancies. For large-scale training, slight differences can accumulate and lead to larger performance deviations.

Regularly check your training metrics and validation results to spot any anomalies early on. If you suspect that certain parameters or settings are not behaving as expected, consider running ablation studies by isolating each factor and training with and without it to measure the exact impact.

Please continue to consult the documentation at https://docs.ultralytics.com for any specifics on options and settings that may help refine your training process. And remember, iteration and experimentation are key in the training of deep learning models.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #19870, which introduces the new CutMix image augmentation feature in Ultralytics 8.3.119. With CutMix, users can now apply an advanced augmentation technique that improves model robustness by simulating realistic occlusions and challenging scenarios during training. This update adds a cutmix option alongside existing methods like mosaic and mixup, updates configuration files, and enhances the documentation with detailed guides and visuals. The integration is seamless, allowing flexible tuning and easy adoption within the training pipeline.

Special thanks to the PR author and contributors for their efforts in bringing this feature to Ultralytics.

To test if this resolves your issue, please update your package with:
```
pip install -U ultralytics>=8.3.119
```
Alternatively, you can install the latest code from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Feel free to wait for the next official PyPI release if you prefer.

Your feedback is appreciated—please let us know if this update resolves the problem or if you encounter any further issues. Thank you for reporting this and helping improve Ultralytics! 🙏
